wagtail/admin/views/generic/base.py

```diff
@@ -113,3 +120,199 @@ def post(self, request, *args, **kwargs):
+            )
+
+        return SearchForm(
+            placeholder=_("Search %(model_name)s")
+            % {"model_name": self.model._meta.verbose_name_plural}
+        )
+
+    def get_queryset(self):
+        queryset = super().get_queryset()
+
+        self.filters, queryset = self.filter_queryset(queryset)
+
+        ordering = self.get_ordering()
+        if ordering:
+            if not isinstance(ordering, (list, tuple)):
+                ordering = (ordering,)
+            queryset = queryset.order_by(*ordering)
+
+        # Preserve the model-level ordering if specified, but fall back on
+        # PK if not (to ensure pagination is consistent)
+        if not queryset.ordered:
+            queryset = queryset.order_by("pk")
+
+        return queryset
+
+    def paginate_queryset(self, queryset, page_size):
+        paginator = self.get_paginator(
+            queryset,
+            page_size,
+            orphans=self.get_paginate_orphans(),
+            allow_empty_first_page=self.get_allow_empty(),
+        )
+
+        page_number = self.request.GET.get(self.page_kwarg)
+        page = paginator.get_page(page_number)
+        return (paginator, page, page.object_list, page.has_other_pages())
+
+    def get_filterset_class(self):
+        if self.filterset_class:
+            return self.filterset_class
+
+        if not self.list_filter or not self.model:
+            return None
+
+        class Meta:
+            model = self.model
+            fields = self.list_filter
+
+        return type(
+            f"{self.model.__name__}FilterSet",
+            (WagtailFilterSet,),
+            {"Meta": Meta},
+        )
+
+    def filter_queryset(self, queryset):
+        # construct filter instance (self.filters) if not created already
+        if self.filterset_class and self.filters is None:
+            self.filters = self.filterset_class(
+                self.request.GET, queryset=queryset, request=self.request
+            )
+            queryset = self.filters.qs
+        elif self.filters:
+            # if filter object was created on a previous filter_queryset call, re-use it
+            queryset = self.filters.filter_queryset(queryset)
+
+        return self.filters, queryset
+
+    def search_queryset(self, queryset):
+        if not self.search_query:
+            return queryset
+
+        if class_is_indexed(queryset.model) and self.search_backend_name:
+            search_backend = get_search_backend(self.search_backend_name)
+            if self.use_autocomplete:
+                return search_backend.autocomplete(
+                    self.search_query, queryset, fields=self.search_fields
+                )
+            return search_backend.search(
+                self.search_query, queryset, fields=self.search_fields
+            )
+
+        filters = {
+            field + "__icontains": self.search_query
+            for field in self.search_fields or []
+        }
+        return queryset.filter(**filters)
```

Comment:
This code was extracted from `ReportView` in 4bedf6e045f6beb01eea1d00077541cffc0fc722, I didn't notice until recently but I'm pretty sure the lookups should've been chained with a `|` (`OR`) operator. Shall we change it now or do you think we should make a separate PR for it?

Reply:
Good spot! I think that's one for a separate PR.

On reflection, I'm not sure that this logic (or indeed the rest of `search_queryset`) really belongs in BaseListingView - since this is meant to cover listings of all kinds, it's not really the right place to embed knowledge about the various mechanisms for searching models (indexed vs non-indexed) - that's more a job for generic.models.IndexView. I guess we want the search box UI to still be handled in BaseListingView, but disabled by default with `is_searchable = False`, and subclasses need to supply their own `search_queryset` method if they want to take advantage of that.